streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java

```diff
@@ -156,24 +134,24 @@ public boolean readyToProcess(final long wallClockTime) {
             final TopicPartition partition = entry.getKey();
             final RecordQueue queue = entry.getValue();
 
-            final Long nullableFetchedLag = fetchedLags.get(partition);
+            final OptionalLong fetchedLag = lagProvider.apply(partition);
```

Comment:
Wearing my paranoid hat here: `readyToProcess` is on the critical path, called per record, while we would only update the underlying lag at most as frequent as the consumer poll rate. And in practice we would fall in to the first condition `!queue.isEmpty()` most of the time. On the other hand, `partitionLag` call on `SubscriptionState` is synchronized and could slow down the fetching thread (well, maybe just a bit). So could we call the provider only necessary, i.e. the queue is empty and the lag is either == 0 or not present?

Reply:
Thanks, this is a good suggestion.